cmake/FindLIBHWLOC.cmake

```diff
@@ -15,24 +15,25 @@ get_filename_component(LIBHWLOC_INCLUDE_DIR ${LIBHWLOC_HEADER} DIRECTORY)
 set(LIBHWLOC_INCLUDE_DIRS ${LIBHWLOC_INCLUDE_DIR})
 
 if(LIBHWLOC_INCLUDE_DIR)
-    if(EXISTS "${LIBHWLOC_INCLUDE_DIR}/hwloc/autogen/config.h")
-        file(
-            STRINGS "${LIBHWLOC_INCLUDE_DIR}/hwloc/autogen/config.h"
-            LIBHWLOC_VERSION
-            REGEX
-                "#define[ \t]HWLOC_VERSION[ \t]\"([0-9]+.[0-9]+.[0-9]+(rc[0-9])?)(-git)?\""
-        )
-        string(
-            REGEX
-            REPLACE
-                "#define[ \t]HWLOC_VERSION[ \t]\"([0-9]+.[0-9]+.[0-9]+(rc[0-9])?)(-git)?\""
-                "\\1"
-                LIBHWLOC_VERSION
-                "${LIBHWLOC_VERSION}")
+    if(EXISTS "${LIBHWLOC_INCLUDE_DIR}/hwloc.h")
+        file(STRINGS "${LIBHWLOC_INCLUDE_DIR}/hwloc.h" LIBHWLOC_API_VERSION
+             REGEX "#define[ \t]HWLOC_API_VERSION[ \t]0x[0-9]+")
+        string(REGEX
+               REPLACE "#define[ \t]HWLOC_API_VERSION[ \t](0x[0-9]+)" "\\1"
+                       LIBHWLOC_API_VERSION "${LIBHWLOC_API_VERSION}")
+
+        math(EXPR LIBHWLOC_API_PATCH "(${LIBHWLOC_API_VERSION} & 0xFF)")
+        math(EXPR LIBHWLOC_API_MINOR "(${LIBHWLOC_API_VERSION} & 0xFF00) >> 8")
+        math(EXPR LIBHWLOC_API_MAJOR
+             "(${LIBHWLOC_API_VERSION} & 0xFF0000) >> 16")
+
+        set(LIBHWLOC_API_VERSION
+            "${LIBHWLOC_API_MAJOR}.${LIBHWLOC_API_MINOR}.${LIBHWLOC_API_PATCH}")
+
```

Comment:
NIT: parsing c header in cmake by using regex is a red flag for me - we could just create C program to extract this information - i.e like one bellow. I'm not fully convinced that this is better - but regex and math inside of cmake also are not to great.

Following code is AI generated and might not work. It might be better to just create support tool and compile it normally instead of using try_run - but i'm leaving this to your consideration.

```suggestion
    # Define the C code that accepts an argument for the version part
    set(HWLOC_VERSION_CODE "
#include <hwloc.h>
#include <stdlib.h>
int main(int argc, char* argv[]) {
    int part = atoi(argv[1]); // Convert the argument to an integer
    if(part == 0) {
        return (HWLOC_API_VERSION & 0xFF0000) >> 16; // Major version
    } else if(part == 1) {
        return (HWLOC_API_VERSION & 0xFF00) >> 8; // Minor version
    }
    return (HWLOC_API_VERSION & 0xFF); // Patch version    
}")
    foreach(idx RANGE 0 2)
        try_run(RUN_RESULT_VAR COMPILE_RESULT_VAR
                ${CMAKE_BINARY_DIR}
                "${HWLOC_VERSION_CODE}"
                ARGS ${idx}
                CMAKE_FLAGS "-DINCLUDE_DIRECTORIES=${LIBHWLOC_INCLUDE_DIR}"
                COMPILE_OUTPUT_VARIABLE COMPILE_OUTPUT
                RUN_OUTPUT_VARIABLE VERSION_OUTPUT)
        if(${idx} EQUAL 0)
            set(LIBHWLOC_API_MAJOR ${VERSION_OUTPUT})
        elseif(${idx} EQUAL 1)
            set(LIBHWLOC_API_MINOR ${VERSION_OUTPUT})
        else()
            set(LIBHWLOC_API_PATCH ${VERSION_OUTPUT})
        endif()
    endforeach()

    set(LIBHWLOC_API_VERSION "${LIBHWLOC_API_MAJOR}.${LIBHWLOC_API_MINOR}.${LIBHWLOC_API_PATCH}")

```

an approach with separate program (also AI generated not tested)

``` c
#include <stdio.h>
#include <stdlib.h>
#include "hwloc.h"

int main(int argc, char** argv) {
    int major = (HWLOC_API_VERSION & 0xFF0000) >> 16;
    int minor = (HWLOC_API_VERSION & 0xFF00) >> 8;
    int patch = (HWLOC_API_VERSION & 0xFF);

    int component = atoi(argv[1]); // Convert the argument to an integer
    if(component == 0) {
        return major;
    } else if(component == 1) {
        return minor;
    } else {
        return patch;
    }
}
```
``` cmake
if(EXISTS "${LIBHWLOC_INCLUDE_DIR}/hwloc.h")
    # Create an executable for extracting the version
    add_executable(hwloc_version_extractor hwloc_version_extractor.c)
    target_include_directories(hwloc_version_extractor PRIVATE ${LIBHWLOC_INCLUDE_DIR})

    # Execute the program for each version component
    foreach(version_idx IN ITEMS 0 1 2)
        execute_process(COMMAND hwloc_version_extractor ${version_idx}
                        RESULT_VARIABLE result
                        OUTPUT_VARIABLE output_var
                        ERROR_QUIET
                        OUTPUT_STRIP_TRAILING_WHITESPACE)

        # Assigning the exit code directly to the version component variable
        if(version_idx EQUAL 0)
            set(LIBHWLOC_API_MAJOR ${result})
        elseif(version_idx EQUAL 1)
            set(LIBHWLOC_API_MINOR ${result})
        else()
            set(LIBHWLOC_API_PATCH ${result})
        endif()
    endforeach()

    # Combine the version components
    set(LIBHWLOC_API_VERSION "${LIBHWLOC_API_MAJOR}.${LIBHWLOC_API_MINOR}.${LIBHWLOC_API_PATCH}")
endif()
```